### PR TITLE
Validate digest length on parsing

### PR DESCRIPTION
**Pull request by tonistiigi**
@aaronlehmann @dmcgowan 

The new error is mostly so that `digest.Set` doesn't need a special way to figure out if invalid digest had a algorithm prefix or not.

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

### CODE REVIEW

**Comment by stevvooe** [comment]
This assumes that digests have fixed size output. Do we want to force this on all future hash implementations? Are there any variable length hashes that we may want to support in the future?

I think the answer is yes, all digest implementations have a fixed size hash output.

**Comment by aaronlehmann** [comment]
I think it's reasonable to assume a fixed hash size.

File: digest/digester.go [hunk 1]
```diff
@@ -54,6 +54,15 @@ func (a Algorithm) String() string {
 	return string(a)
 }
 
+// Size returns number of bytes returned by the hash.
+func (a Algorithm) Size() int {
+	h, ok := algorithms[a]
+	if !ok {
+		return 0
```
Comment (stevvooe) [inline digest/digester.go]: Does this break tarsum?

**Comment by tonistiigi** [comment]
@stevvooe If there is ever a case where fix length doesn't work, `Size` can be easily changed to a `ValidateSize` method with exceptions. But don't see need for that atm.

File: digest/digester.go (hunk 1 above)
Comment (tonistiigi) [inline digest/digester.go]: Tarsum has its own parse method so isn't affected(not better or worse). I can try to fix tarsum as well if needed, of course.

**Comment by stevvooe** [comment]
@tonistiigi I think we're okay. SHA3, blake etc. all have fixed length output. As a matter of security, this is probably more than okay to enforce.

LGTM

**Comment by dmcgowan** [comment]
LGTM

**Pull request merged**